domain/model/state/modelstate.go

```diff
@@ -105,6 +111,402 @@ func (s *ModelState) Delete(ctx context.Context, uuid coremodel.UUID) error {
+		if err != nil {
+			return errors.Errorf("getting model uuid: %w", err)
+		}
+		cons, err = s.getModelConstraints(ctx, modelUUID, tx)
+		if err != nil {
+			return errors.Capture(err)
+		}
+		if cons.UUID == "" {
+			// No constraint exists for the model, no furhter queries are needed.
+			return nil
+		}
+		err = tx.Query(ctx, selectTagStmt, cons).GetAll(&tags)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Errorf("getting constraint tags: %w", err)
+		}
+		err = tx.Query(ctx, selectSpaceStmt, cons).GetAll(&spaces)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Errorf("getting constraint spaces: %w", err)
+		}
+		err = tx.Query(ctx, selectZoneStmt, cons).GetAll(&zones)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Errorf("getting constraint zones: %w", err)
+		}
+		return nil
+	})
+	if err != nil {
+		return constraints.Value{}, errors.Capture(err)
+	}
+	return cons.toValue(tags, spaces, zones)
+}
+
+func (s *ModelState) getModelConstraints(ctx context.Context, modelUUID coremodel.UUID, tx *sqlair.TX) (dbConstraint, error) {
+	stmt, err := s.Prepare(`
+SELECT c.uuid AS &dbConstraint.uuid,
+       c.arch AS &dbConstraint.arch,
+       c.cpu_cores AS &dbConstraint.cpu_cores,
+       c.cpu_power AS &dbConstraint.cpu_power,
+       c.mem AS &dbConstraint.mem,
+       c.root_disk AS &dbConstraint.root_disk,
+       c.root_disk_source AS &dbConstraint.root_disk_source,
+       c.instance_role AS &dbConstraint.instance_role,
+       c.instance_type AS &dbConstraint.instance_type,
+       ct.value AS &dbConstraint.container_type,
+       c.virt_type AS &dbConstraint.virt_type,
+       c.allocate_public_ip AS &dbConstraint.allocate_public_ip,
+       c.image_id AS &dbConstraint.image_id
+FROM   model_constraint mc
+       JOIN "constraint" c ON c.uuid = mc.constraint_uuid
+       JOIN container_type ct ON ct.id = c.container_type_id
+WHERE  mc.model_uuid = $dbModelConstraint.model_uuid
+`, dbConstraint{}, dbModelConstraint{})
+	if err != nil {
+		return dbConstraint{}, errors.Capture(err)
+	}
+
+	modelConstraint := dbModelConstraint{ModelUUID: modelUUID.String()}
+	var constraint dbConstraint
+	err = tx.Query(ctx, stmt, modelConstraint).Get(&constraint)
+	if errors.Is(err, sql.ErrNoRows) {
+		return dbConstraint{}, nil
+	}
+	if err != nil {
+		return dbConstraint{}, errors.Errorf("getting model constraint for model %q: %w", modelUUID, err)
+	}
+	return constraint, nil
+}
+
+// SetModelConstraints sets the model constraints, including tags, spaces, and zones.
+// It returns an error satisfying [networkerrors.SpaceNotFound] if a space to set does not exist,
+// [modelerrors.NotFound] if the model does not exist.
+func (s *ModelState) SetModelConstraints(ctx context.Context, consValue constraints.Value) error {
+	db, err := s.DB()
+	if err != nil {
+		return errors.Capture(err)
+	}
+	if consValue.Container == nil {
+		consValue.Container = &NONEContainerType
```

Comment:
I don't think we should be doing this. This is now how we implement this today and we have a lot of logic that reads these values. Explicitly now None has to be set explicitly and null has a different meaning.

While this modelling is not great we should not be introducing this change into Juju without having a conversation and understand the side affects.